Keep flattened columns as extra columns

sgiehl opened the pull request:
Flattened columns will be added with their dimension id. If dimension id is the same for subtables (like in pages reports) the values will be concatenated.
If a table has no dimension a numerated label will be used instead. e.g. `label1`

fixes #12163

sgiehl committed:
Add reports dimensions to metadata of report and rows

sgiehl committed:
updates test files

sgiehl committed:
translate dimension columns

mattab commented on tests/PHPUnit/System/expected/test_FlattenReports__Referrers.getWebsites_day.xml:
```diff
@@ -34,6 +38,8 @@
 		<sum_visit_length>1</sum_visit_length>
 		<bounce_count>0</bounce_count>
 		<nb_visits_converted>0</nb_visits_converted>
+		<Referrers.Website>www.referrer2.com</Referrers.Website>
```
Invalid XML

sgiehl replied:
actually it's not invalid as `.` is valid for xml tags as long as it doesn't begin with `.`. But I'll replace that with `_`

mattab reviewed (dismissed):
* For the processed report, we may need to add a new system test in case they are not tested yet

sgiehl commented:
@mattab how should processed reports be handled?
Currently those additional columns aren't included as they are not defined in the reports metadata.
Tweaking the reports metadata by default is no option, as those columns aren't included in non-flat reports. So we would need to tweak the report results when the report is being processed, but I'm currently a bit lost how to do that proper. 
Flat processed reports already have an incorrect label, as the dimension defined in metadata is used. Instead the label is combined with the one of the subtables. For example https://demo.piwik.org/index.php?module=API&method=API.getProcessedReport&apiModule=Referrers&apiAction=getSearchEngines&format=XML&idSite=3&period=month&date=2017-10-14&flat=1&filter_limit=25 uses `Search Engine` as label. Instead it would need to be `Search Engine - Keyword`. But actually there is no information that the keyword was merged in the report.

sgiehl commented:
@mattab @tsteur anyone able to give me a smart hint how to proceed?

tsteur commented:
True, we would need to make sure to manipulate the `label` as well when flatten is used. Might need to post an event like `API.getProcessedReport.end` in `ProcessedReport::getProcessedReport` or handle it directly in there (would need to forward flatten parameter). In `Report::buildReportMetadata()` we cannot really do it as we don't know there whether a report is flattened or not (and we cannot rely on `$_GET/$_POST`).

I totally see what you mean... 

I wonder if we can in `Report::buildReportMetadata()` always return all labels of all possible dimensions like

```php
        $report['dimensions']           = [$this->getDimension()->getId() => $this->getDimension()->getName(), $this->getSubtableDimension()->getId() => $this->getSubtableDimension()->getName()]; // note subtable dimensions might be not defined, and some reports have multiple dimensions so we would need a method like `$this->getDimensions()` or so.
        $report['metrics']              = $this->getMetrics();
        $report['metricsDocumentation'] = $this->getMetricsDocumentation();
        $report['processedMetrics']     = $this->getProcessedMetrics();
```

So we basically add this always to the report metadata output. And maybe then in `ProcessedReport::getProcessedReport()` we can do like an `implode(' - ', $report['dimensions'])` when flatten is used for the label? The ` - `might be needed to be adjusted as a different string can be configured in the report class or so.

We would also need to add the information to `columns => $columns` in the processed report class. 

I just checked Piwik Mobile roughly and it will likely not break it, but needs to be double checked. Actually, I just checked again and there is quite a chance it might break Piwik Mobile as it will think that any column specified in `columns` is a metric.  

Likely this issue can break various implementations because. Adding something to columns that is not a metric can lead to breaking behaviour, and adding new row columns that are not metrics can break behaviour as well as so far it is assumed that numbers are returned in all fields except `label` which will be no longer the case. Wondering if better to do in Piwik 4?

mattab commented:
> Wondering if better to do in Piwik 4?

Would be great to do it before in Processed Report output :+1:

tsteur commented:
In report metadata could work to not break API. In Piwik 4 we would then maybe actually put it into the row element (or maybe not, to be discussed).

mattab commented:
@sgiehl Any update on this one? it is an important one for Custom Reports users

sgiehl commented:
I've updated the PR.
Everything is now added to the metadata.
If the report has more than one dimension a new metadata entry `dimensions` will be added, listing all dimensions. (Are report is considered as having more dimension if `getDimensions` returns an array with more than one entries, by default the array contains the reports dimension and if available the dimension of the subtable)
For flattened reports those dimensions will be additionally added to each rows metadata. This also applies for processed reports as row metadata is taken from all sub reports.

mattab commented:
Looks good to me and should be safe and keep BC :+1: 

Feedback:

* for the dimension ID in the API output, could we use the Segment string eg. `<downloadUrl>...` instead of `<Actions_DownloadUrl>...`, eg. `<dimension3>` instead of `<CustomDimensions_dimension3>`
  * (to indicate the column is part of the label/main dimension, we could maybe even name `label_downloadUrl` or `label_dimension3`? not too sure if it's better)
* as expected there are some tests failing in CustomDimensions and MarketingCampaignReporting plugins

sgiehl commented:
> for the dimension ID in the API output, could we use the Segment string eg. <downloadUrl>... instead of <Actions_DownloadUrl>..., eg. <dimension3> instead of <CustomDimensions_dimension3>
(to indicate the column is part of the label/main dimension, we could maybe even name label_downloadUrl or label_dimension3? not too sure if it's better)

Using the segment name can only work if a segment is defined. As this is a generic solution there might be plugins where no segment is defined for a subtable. In this case we would need to have a fallback

mattab commented:
Good point. When there is no segment defined, maybe we could fallback to `$Plugin_$Dimension`? Btw, was this 'dimension key' already used somewhere else in Piwik?

sgiehl committed:
fix possible error when no report is available

sgiehl commented:
It currently uses the dimension id, so probably something that might be used somewhere else in the future. So I wouldn't rename it to anything else. Also this label name needs to be used to add the translations...

diosmosis commented on core/API/DataTableManipulator/Flattener.php:
```diff
@@ -137,7 +156,18 @@ private function flattenRow(Row $row, $rowId, DataTable $dataTable,
+
+            $report           = ReportsProvider::factory($this->apiModule, $this->getApiMethodForSubtable($this->request));
+            $subDimension     = $report->getDimension();
+            $subDimensionName = $subDimension ? str_replace('.', '_', $subDimension->getId()) : 'label' . (substr_count($prefix, $this->recursiveLabelSeparator) + 1);
```
Is it possible to have a test for when a report has no dimension or subdimension? To trigger the labelN logic.

diosmosis commented on tests/PHPUnit/System/expected/test_CustomEvents_Events.getAction_flat__API.getProcessedReport_day.xml:
```diff
@@ -404,8 +408,163 @@
+				<Events_EventAction>Search</Events_EventAction>
+				<Events_EventName>Search query here</Events_EventName>
+			</row>
+		</result>
```
I see now there will be rows w/ dimensions even if there is no data. Will this be a BC break? I wonder if the mobile app or other types of apps will experience different behavior because of this.

diosmosis replied:
cc @tsteur in case he has some insight

diosmosis commented:
Tested locally & looks good, noticed one possible BC issue (commented).

Also, I think the tests don't cover processed reports when there is data and they don't cover the labelN logic. I think when those are added would be good to merge.

diosmosis commented:
@sgiehl did some more testing & noticed this doesn't seem to work w/ CustomReports:

![image](https://user-images.githubusercontent.com/125140/43176792-2d57ca12-8f7a-11e8-8059-96f9a06dc129.png)

that's w/ these dimensions:

![image](https://user-images.githubusercontent.com/125140/43176807-3bc44422-8f7a-11e8-9cad-a20fed4729ab.png)

mattab reviewed (changes requested):
@sgiehl can you please investigate Benaka's feedback
*  tests don't cover processed reports when there is data and they don't cover the labelN logic. can you please add those?
* test this is working for custom reports (which is where having multiple dimensions as columns will be awesome)
* is it expected that this test output changes? Will this affect BC? https://github.com/matomo-org/matomo/pull/12199/files/fe2cb5a7f1b1ed404f86c391247982c3f18e99e9#diff-5c16d2fae8e978c9508dc0f4e4368a1b
* some other system and tests are still failing?

sgiehl committed:
update tests

sgiehl commented:
> test this is working for custom reports (which is where having multiple dimensions as columns will be awesome)

I've investigated the custom reports plugin together with this branch now for a while and figured out where the problem is. The Report classes for CustomReports are created dynamically the according dimension is detected by checking if it's a subtable report or not (check for the get parameter idSubtable). As this parameter isn't set when flattening the report ([here](https://github.com/matomo-org/matomo/blob/d35c198ca215a8b5ba3be134acbbbaf2a9ecb51e/core/API/DataTableManipulator/Flattener.php#L160)) it always returns the dimension of the main report, which causes the flattening to simply concatenate the value instead of adding additional dimension entries. We could temporarily set the get parameter at that point or extend the `ReportsProvider::factory` to receive an additional subTableId or similar.

sgiehl committed:
Improve subdimension detection

sgiehl committed:
Adjust tests for labelX logic

sgiehl commented:
@diosmosis @mattab 

> tests don't cover processed reports when there is data and they don't cover the labelN logic. can you please add those?

Adjusted a test to cover that.

> test this is working for custom reports (which is where having multiple dimensions as columns will be awesome)

With latest changes it works with Custom Reports, but only with reports having 2 Dimensions only. For supporting 3 dimensions we may need to adjust a bit more. See my comment before.

> is it expected that this test output changes? Will this affect BC? https://github.com/matomo-org/matomo/pull/12199/files/fe2cb5a7f1b1ed404f86c391247982c3f18e99e9#diff-5c16d2fae8e978c9508dc0f4e4368a1b

The Link doesn't work anymore, but guess it's related to the added metadata. That is expected behaviour.

> some other system and tests are still failing?

Remaining failing tests should be for CustomDimensions and MarketingCampaignsReporting plugin only and need to be adjusted when ready to merge...

mattab commented:
It is important to support custom reports with 3 dimensions and also have system tests for this. I'm not sure which solution is best though among the two you suggest in your comment..

sgiehl commented:
@mattab  After doing some more testing I guess both methods won't work for 3 Dimensions. To be honest I'm a bit stuck in finding a proper solution how to handle the dynamically added reports by CustomReports plugin without having to refactor the whole Report stuff in Matomo. Everything else that comes up to my mind would be bit too hackish...

tsteur commented:
Was the report class problem maybe fixed recently? I think the cache was made per idSite etc. Not sure if that helps?

sgiehl commented:
@tsteur  no the problem is, that there is no simple way to support reports of CustomDimensions plugin that have 3 dimensions. The report flattener fails to get the correct dimension for the subtables as the ReportFactory returns the main report.

tsteur commented:
@sgiehl I think I've managed to add the third dimension but not sure if there is any other problem? I noticed the first dimension wasn't really shown... not sure why... we could make the methods `getThirdLeveltableDimension` and even `getSecondLeveltableDimension()` API in `Report` class maybe as it is more clear than just `getSubtableDimension()`

```diff
diff --git a/core/API/DataTableManipulator/Flattener.php b/core/API/DataTableManipulator/Flattener.php
index e1bb7023ad..342480d054 100644
--- a/core/API/DataTableManipulator/Flattener.php
+++ b/core/API/DataTableManipulator/Flattener.php
@@ -75,7 +75,7 @@ class Flattener extends DataTableManipulator
 
         $dimensionName = !empty($dimension) ? str_replace('.', '_', $dimension->getId()) : 'label1';
 
-        $this->flattenDataTableInto($dataTable, $newDataTable, $dimensionName);
+        $this->flattenDataTableInto($dataTable, $newDataTable, $level = 1, $dimensionName);
 
         return $newDataTable;
     }
@@ -85,10 +85,10 @@ class Flattener extends DataTableManipulator
      * @param $newDataTable
      * @param $dimensionName
      */
-    protected function flattenDataTableInto($dataTable, $newDataTable, $dimensionName, $prefix = '', $logo = false)
+    protected function flattenDataTableInto($dataTable, $newDataTable, $level, $dimensionName, $prefix = '', $logo = false)
     {
         foreach ($dataTable->getRows() as $rowId => $row) {
-            $this->flattenRow($row, $rowId, $newDataTable, $dimensionName, $prefix, $logo);
+            $this->flattenRow($row, $rowId, $newDataTable, $level, $dimensionName, $prefix, $logo);
         }
     }
 
@@ -99,7 +99,8 @@ class Flattener extends DataTableManipulator
      * @param string $dimensionName
      * @param bool $parentLogo
      */
-    private function flattenRow(Row $row, $rowId, DataTable $dataTable, $dimensionName,
+    private function flattenRow
+    (Row $row, $rowId, DataTable $dataTable, $level, $dimensionName,
                                 $labelPrefix = '', $parentLogo = false)
     {
         $origLabel = $label = $row->getColumn('label');
@@ -162,6 +163,10 @@ class Flattener extends DataTableManipulator
                 $subDimension = $report->getSubtableDimension();
             }
 
+            if ($level === 2 && method_exists($report, 'getThirdLeveltableDimension')) {
+                $subDimension = $report->getThirdLeveltableDimension();
+            }
+
             if (empty($subDimension)) {
                 $report           = ReportsProvider::factory($this->apiModule, $this->getApiMethodForSubtable($this->request));
                 $subDimension     = $report->getDimension();
@@ -175,7 +180,7 @@ class Flattener extends DataTableManipulator
                 }
             }
 
-            $this->flattenDataTableInto($subTable, $dataTable, $subDimensionName, $prefix, $logo);
+            $this->flattenDataTableInto($subTable, $dataTable, $level + 1, $subDimensionName, $prefix, $logo);
         }
     }

```

sgiehl committed:
Makes flattener compatible with 3 dimensions

sgiehl commented:
@tsteur  thanks, that helped. Needed to adjust some more stuff so all dimensions are added correctly. But now it looks better.
What about adding the new methods as API? Do you think it makes sense, or is the usage of custom reports plugin too "special"?

tsteur commented:
I think it would make sense to add those methods to Report class as API especially since we might also add reports to core with 3 dimensions maybe.

sgiehl committed:
Adds new method getThirdLeveltableDimension to report class

tsteur committed:
Do not ask for 2fa authentication code when CoreUpdater is being requested (#13796)

Could fix an edge case where user is logged in, but hasn't confirmed the auth code (so the user is not actually logged in), and then an update appears.

ozdemirburak committed:
Added Fallback Method for Alexa in SEO Plugin (#13552)

* added fallback method for Alexa, fixes issue #13427

* do not use short array syntax for consistency with other methods

* use mini link for Alexa, use DomXPath to filter out the global ranking instead of regex

tsteur committed:
Use db sessions by default, deprecate file session handler (#13540)

* use db sessions by default, deprecate file session handler

* trying to fix tests

tsteur committed:
Prevent trigger errors on demand for instances that are opened to anonymous (#13535)

fix #13513

tsteur committed:
Remove the previous exception in base validator so the same error is not printed twice (#13801)

diosmosis committed:
Fixing build  (#13802)

* update submodule

* Update screenshots and try to get test to pass.

diosmosis committed:
Get SingleMetricView to pass. (#13803)

tsteur committed:
Quickform2 throws warnings with PHP7.2 (#13463)

fixes #13272

Haven't actually tested it but should fix the issue. If tests pass, the logic would be still the same. I don't have a PHP 7.2 running here otherwise at the moment

tsteur committed:
Send bulk requests in chunks when needed (#13444)

* send bulk requests in chunks

* send requests correctly

tsteur committed:
Make log and report data screen less technical (#13464)

tsteur committed:
When you are logged out, the URL gets lost when you log in (#13441)

It won't remember any hash as the hash won't be visible in the referrer etc but it would work for most other pages.

To make it work for hash it would get likely way more complicated like we would need to persist it through JS, temporarily store it somewhere and redirect accordingly. It fixes the case mentioned in the issue.

fix https://github.com/matomo-org/matomo/issues/13328

tsteur committed:
show full information of URL only on extra click (#13585)

tsteur committed:
Add option to opt in to use send beacon (#13451)

* Add option to opt in to use send beacon

* Fix JS tracker test.

diosmosis commented:
Tests are failing, looks like the `<url>` field might be incorrect in some reports (like getSearchEngines).

sgiehl committed:
do not overrwite existing subrow metadata

sgiehl committed:
update test files

sgiehl commented:
slightly updated the PR to fix the url fields

diosmosis committed:
Merge branch '3.x-dev' into flattenexport
